tests/test_geometry_2d.h

```diff
@@ -0,0 +1,503 @@
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef TEST_GEOMETRY_2D_H
+#define TEST_GEOMETRY_2D_H
+
+#include "core/math/geometry_2d.h"
+#include "core/templates/vector.h"
+
+#include "thirdparty/doctest/doctest.h"
+
+namespace TestGeometry2D {
+
+TEST_CASE("[Geometry2D] Point in circle") {
+	CHECK(Geometry2D::is_point_in_circle(Vector2(0, 0), Vector2(0, 0), 1.0));
+
+	CHECK(Geometry2D::is_point_in_circle(Vector2(0, 0), Vector2(11.99, 0), 12));
+	CHECK(Geometry2D::is_point_in_circle(Vector2(-11.99, 0), Vector2(0, 0), 12));
+
+	CHECK_FALSE(Geometry2D::is_point_in_circle(Vector2(0, 0), Vector2(12.01, 0), 12));
+	CHECK_FALSE(Geometry2D::is_point_in_circle(Vector2(-12.01, 0), Vector2(0, 0), 12));
+
+	CHECK(Geometry2D::is_point_in_circle(Vector2(7, -42), Vector2(4, -40), 3.7));
+	CHECK_FALSE(Geometry2D::is_point_in_circle(Vector2(7, -42), Vector2(4, -40), 3.5));
+}
+
+TEST_CASE("[Geometry2D] Point in triangle") {
+	CHECK(Geometry2D::is_point_in_triangle(Vector2(0, 0), Vector2(-1, 1), Vector2(0, -1), Vector2(1, 1)));
+	CHECK_FALSE(Geometry2D::is_point_in_triangle(Vector2(-1.01, 1.0), Vector2(-1, 1), Vector2(0, -1), Vector2(1, 1)));
+
+	CHECK(Geometry2D::is_point_in_triangle(Vector2(3, 2.5), Vector2(1, 4), Vector2(3, 2), Vector2(5, 4)));
+	CHECK(Geometry2D::is_point_in_triangle(Vector2(-3, -2.5), Vector2(-1, -4), Vector2(-3, -2), Vector2(-5, -4)));
+	CHECK_FALSE(Geometry2D::is_point_in_triangle(Vector2(0, 0), Vector2(1, 4), Vector2(3, 2), Vector2(5, 4)));
+}
+
+TEST_CASE("[Geometry2D] Point in polygon") {
+	Vector<Vector2> p;
+	CHECK_FALSE(Geometry2D::is_point_in_polygon(Vector2(0, 0), p));
+
+	p.push_back(Vector2(-88, 120));
+	p.push_back(Vector2(-74, -38));
+	p.push_back(Vector2(135, -145));
+	p.push_back(Vector2(425, 70));
+	p.push_back(Vector2(68, 112));
+	p.push_back(Vector2(-120, 370));
+	p.push_back(Vector2(-323, -145));
+	CHECK_FALSE(Geometry2D::is_point_in_polygon(Vector2(-350, 0), p));
+	CHECK_FALSE(Geometry2D::is_point_in_polygon(Vector2(-110, 60), p));
+	CHECK_FALSE(Geometry2D::is_point_in_polygon(Vector2(412, 96), p));
+	CHECK_FALSE(Geometry2D::is_point_in_polygon(Vector2(83, 130), p));
+	CHECK_FALSE(Geometry2D::is_point_in_polygon(Vector2(-320, -153), p));
+
+	CHECK(Geometry2D::is_point_in_polygon(Vector2(0, 0), p));
+	CHECK(Geometry2D::is_point_in_polygon(Vector2(-230, 0), p));
+	CHECK(Geometry2D::is_point_in_polygon(Vector2(130, -110), p));
+	CHECK(Geometry2D::is_point_in_polygon(Vector2(370, 55), p));
+	CHECK(Geometry2D::is_point_in_polygon(Vector2(-160, 190), p));
```

Comment:
I'd also add a test case with a point being exactly on polygon's edge/point, some implementations distinguish those cases, so I wonder whether Godot treats a point exactly on polygon's boundary as *outside*.

If those test cases already cover my suggestion, then I'd add comments to them or even convert those into dedicated test cases (or [subcases](https://github.com/onqtam/doctest/blob/master/doc/markdown/tutorial.md#test-cases-and-subcases)).

Reply:
Good point. I don't think i included such cases here but i remember trying this in a test project and i think it's treated is being inside the polygon. I'm going to add similar cases to the other methodes (`is_point_in_triangle` etc.) as well.

Reply:
I should probably also convert the different test scenarios for the Boolean Logic methods (`intersect_polygons` etc.) to subcases (didn't know they existetd).

Reply:
Yes, I think it's worth it if the test data is the same, but may also add extra indentation which may worsen readability.

But I don't know what happens if a subcase fails (fails the entire test case or just the subcase itself).

Reply:
Yeah, one polygon pretty much never changes. I think it actually improves readability because you can clearly see that the code belongs together to test a certain functionality.

That's from the `SUBCASE` docs:
> If a CHECK() fails - the test is marked as failed but the execution continues - but if a REQUIRE() fails - execution of the test stops

So if a subcase fails, the whole test case fails but execution continues, just like now.